### PR TITLE
The Chronicles of Nania

### PR DESCRIPTION
**Pull request by michael-lutz**

### CODE REVIEW

File: ksim/task/ppo.py [hunk 1]
```diff
@@ -40,11 +43,29 @@ class PPOConfig(RLConfig):
 
     # General training parameters
     # TODO: none of these except `max_grad_norm` are actually used in the training script
```
Comment (budzianowski) [inline ksim/task/ppo.py]: remove above

File: ksim/task/ppo.py [hunk 2]
```diff
@@ -40,11 +43,29 @@ class PPOConfig(RLConfig):
...
-    learning_rate: float = xax.field(value=1e-3, help="Learning rate for PPO.")
+    learning_rate: float = xax.field(value=1e-4, help="Learning rate for PPO.")
     schedule: str = xax.field(value="adaptive", help="Learning rate schedule 'fixed' | 'adaptive'")
     desired_kl: float = xax.field(value=0.01, help="Desired KL divergence for adaptive LR.")
```
Comment (budzianowski) [inline ksim/task/ppo.py]: remove not implemented

File: ksim/model/formulations.py [hunk 3]
```diff
@@ -96,33 +98,131 @@ def sample_and_log_prob(
...
+                f"obs_std_{obs_name}",
+                nn.initializers.ones,
+                key=(),
+                shape=obs_vec.shape[-1],
+            )
+            for obs_name, obs_vec in obs.items()
+        }
+        # note: initialized here once, will be updated in the training loop
+
+        # do normaliaztion on inputs
```
Comment (budzianowski) [inline ksim/model/formulations.py]:
```suggestion
        # do normalization on inputs
```

File: ksim/model/formulations.py [hunk 4]
```diff
@@ -96,33 +98,131 @@ def sample_and_log_prob(
...
+        normalized_obs_dict = {
+            obs_name: (obs_vec - obs_mean[obs_name].value) / obs_std[obs_name].value
+            for obs_name, obs_vec in obs.items()
+        }
+
+        normalized_obs: FrozenDict[str, Array] = FrozenDict(normalized_obs_dict)
+
+        return normalized_obs
+
+    def normalize_returns(self, returns: Array) -> Array:
```
Comment (budzianowski) [inline ksim/model/formulations.py]: where is this used?

File: ksim/task/ppo.py [hunk 5]
```diff
@@ -175,25 +241,41 @@ def compute_ppo_loss(
...
 
         # get the state-value estimates
-        values = self.apply_critic(model, params, env_state_batch.obs, env_state_batch.command)
+        values = self.apply_critic(model, variables, env_state_batch.obs, env_state_batch.command)
         assert isinstance(values, Array)
         values = values.squeeze(axis=-1)  # values is (time, env)
 
+        advantages = rollout_time_loss_components.advantages
+        if self.config.normalize_advantage:  # TODO: use EWMA for this
+            advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
```
Comment (budzianowski) [inline ksim/task/ppo.py]:
```suggestion
            advantages = (advantages - advantages.mean()) / (advantages.std() + EPSILON)
```

File: ksim/builders/rewards.py [hunk 6]
```diff
@@ -98,6 +98,27 @@ def __call__(self, data: BuilderData) -> T:
...
+        self,
+        action_t_minus_1: Array | None,
+        mjx_data_t: mjx.Data,
+        command_t: FrozenDict[str, Array],
+        action_t: Array,
+        mjx_data_t_plus_1: mjx.Data,
+    ) -> Array:
+        height = mjx_data_t_plus_1.qpos[2]
+        reward = jnp.exp(-jnp.abs(height - self.height_target) * 50)
+        # jax.debug.print("height: {height}", height=height)
```
Comment (budzianowski) [inline ksim/builders/rewards.py]: remove

File: ksim/task/ppo.py [hunk 7]
```diff
@@ -40,11 +43,29 @@ class PPOConfig(RLConfig):
...
     # General training parameters
     # TODO: none of these except `max_grad_norm` are actually used in the training script
-    learning_rate: float = xax.field(value=1e-3, help="Learning rate for PPO.")
+    learning_rate: float = xax.field(value=1e-4, help="Learning rate for PPO.")
     schedule: str = xax.field(value="adaptive", help="Learning rate schedule 'fixed' | 'adaptive'")
     desired_kl: float = xax.field(value=0.01, help="Desired KL divergence for adaptive LR.")
     max_grad_norm: float = xax.field(value=1.0, help="Maximum gradient norm for clipping.")
 
+    # Normalization parameters
+    returns_norm_alpha: float = xax.field(
```
Comment (budzianowski) [inline ksim/task/ppo.py]: If i read it correctly, the `returns_norm_alpha=0.0`makes it look it is not applied but we just stick with initial mean which is even worse?

File: ksim/task/ppo.py [hunk 8]
```diff
@@ -126,6 +167,29 @@ def get_optimizer(self) -> optax.GradientTransformation:
             optax.adam(self.config.learning_rate),
         )
 
+    @legit_jit(static_argnames=["self", "initial_step"])
+    def update_input_normalization_stats(
```
Comment (budzianowski) [inline ksim/task/ppo.py:178]: I don't think that the normalization should be applied to all envs. Can we make it optional with default=True.

File: ksim/task/ppo.py (hunk 7 above)
Comment (michael-lutz) [reply ksim/task/ppo.py]: Yup, 0.0 means we just normalize with the first rollout and don't adjust afterwards. I think this roughly makes sense since it at least gets the scale right for some larger observation values. Haven't been able to run enough experiments to get a sense of how online updates to normalization affects training.

File: ksim/task/ppo.py (hunk 8 above)
Comment (michael-lutz) [reply ksim/task/ppo.py:178]: Sure, they're initialized to be zero mean and 1 std, so I'll just update in PPO. Parametrizing at the top of PPO. Keeping normalize obs to default true since I think that really helps

File: ksim/task/ppo.py (hunk 2 above)
Comment (michael-lutz) [reply ksim/task/ppo.py]: The optimizer uses it?

Comment (budzianowski) [reply ksim/task/ppo.py]: Apologies, I should have been more explicit - remove `desired_kl` and `schedule` and `# TODO: none of these except `max_grad_norm` are actually used in the training script`

File: ksim/model/formulations.py (hunk 4 above)
Comment (michael-lutz) [reply ksim/model/formulations.py]: Not used, added logic to `ppo.py` directly if the config specifies to normalize returns

**Review by budzianowski — APPROVED** [review]
